### PR TITLE
feat(uri): add `user_info()` API

### LINKED ISSUE #805: Add a `[[nodiscard]] auto URI::userinfo() const -> std::optional<std::string_view>` method

### PR DESCRIPTION
**Pull request by tony-go**
Fixes #805

### CODE REVIEW

File: test/uri/uri_user_info_test.cc [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  EXPECT_EQ(uri.user_info().value(), "user:");
+}
+
+// Note: Using the format "user:password" is depreacted.
+// But we would like to support it for users that would still use it.
+TEST(URI_user_info, depreacted_example) {
+  sourcemeta::jsontoolkit::URI uri{"http://user:password@host:80/"};
+  EXPECT_TRUE(uri.user_info().has_value());
+  EXPECT_EQ(uri.user_info().value(), "user:password");
+}
```
Comment (tony-go) [inline test/uri/uri_user_info_test.cc:17]: I wonder if we should not be strict on this? or at least print a warning.

Comment (jviotti) [reply test/uri/uri_user_info_test.cc:17]: I think it's fine. Even if it's deprecated, is part of the RFC. The warning you put in the docs should be enough!

File: test/uri/uri_user_info_test.cc [hunk 2]
```diff
@@ -0,0 +1,22 @@
+
+#include <gtest/gtest.h>
+#include <sourcemeta/jsontoolkit/uri.h>
+
+TEST(URI_user_info, working_example) {
```
Comment (jviotti) [inline test/uri/uri_user_info_test.cc:5]: Can you test user info with a URN? (if it even makes sense there?)

Comment (jviotti) [reply test/uri/uri_user_info_test.cc:5]: Also maybe test it with a relative URI as well, just for extra coverage, which might come handy when we do our own parser

Comment (tony-go) [reply test/uri/uri_user_info_test.cc:5]: > Can you test user info with a URN? (if it even makes sense there?)

Regarding my first reading on URNs and tags, it seems userinfo does not have to deal with.

Comment (tony-go) [reply test/uri/uri_user_info_test.cc:5]: > lso maybe test it with a relative URI as well

Relative URI represents only path right? I don't think we could use user_info, scheme or port in this case.

Comment (jviotti) [reply test/uri/uri_user_info_test.cc:5]: On relative URIs, exactly! You can to assert that the userinfo is empty

Comment (jviotti) [reply test/uri/uri_user_info_test.cc:5]: > Regarding my first reading on URNs and tags, it seems userinfo does not have to deal with.

yes, I confirm! Maybe still add a test that checks that for a URN, the userinfo returns a non-value?

File: src/uri/include/sourcemeta/jsontoolkit/uri.h [hunk 3]
```diff
@@ -288,8 +288,25 @@ class SOURCEMETA_JSONTOOLKIT_URI_EXPORT URI {
...
+  ///
+  /// const sourcemeta::jsontoolkit::URI uri{"https://user:@host"};
+  /// assert(uri.user_info().has_value());
+  /// assert(uri.user_info().value() == "user:);
+  /// ```
+  ///
+  /// As mentioned in RFC 3986, the format "user:password" is deprecated.
+  /// Applications should not render as clear text any data after the first
+  /// colon. See https://tools.ietf.org/html/rfc3986#section-3.2.1
+  [[nodiscard]] auto user_info() const -> std::optional<std::string_view>;
```
Comment (jviotti) [inline src/uri/include/sourcemeta/jsontoolkit/uri.h:308]: Can we call this `userinfo`? The RFC refers to it as such, even though it uses hyphens for other parts of the grammar. Maybe worth sticking with that convention?

Comment (jviotti) [reply src/uri/include/sourcemeta/jsontoolkit/uri.h:308]: Hm, I don't have strong opinions actually. This looks good too. I don't know

**Review by jviotti — APPROVED** [review]
I left some minor remaining comments but looks good. Feel free to merge when ready!

File: src/uri/include/sourcemeta/jsontoolkit/uri.h (hunk 3 above)
Comment (tony-go) [reply src/uri/include/sourcemeta/jsontoolkit/uri.h]: done

**Pull request merged**